gunpowder/jax/nodes/train.py

```diff
@@ -0,0 +1,288 @@
+        spawn_subprocess: bool = False,
+        n_devices: Optional[int] = None
+    ):
+
+        # not yet implemented
+        gradients = gradients
+
+        super(Train, self).__init__(
+            inputs, outputs, gradients, array_specs,
+            spawn_subprocess=spawn_subprocess
+        )
+
+        self.model = model
+        self.checkpoint_basename = checkpoint_basename
+        self.save_every = save_every
+        if n_devices is None:
+            n_devices = jax.local_device_count()  # autodetect available GPUs
+        self.n_devices = n_devices
+        self.local_devices = jax.local_devices()
+        self.keep_n_checkpoints = keep_n_checkpoints
+
+        self.iteration = 0
+
+        if not isinstance(tensorboardX, NoSuchModule) and log_dir is not None:
+            self.summary_writer = tensorboardX.SummaryWriter(log_dir)
+            self.log_every = log_every
+        else:
+            self.summary_writer = None
+            if log_dir is not None:
+                logger.warning("log_dir given, but tensorboardX is not installed")
+
+        self.intermediate_layers = {}
+
+    def replicate_params(self, params):
+        return jax.tree_map(lambda x: jnp.array([x] * self.n_devices), params)
+
+    def start(self):
+        checkpoint, self.iteration = self._get_latest_checkpoint(
+            self.checkpoint_basename)
+
+        if checkpoint is not None:
+
+            logger.info("Resuming training from iteration %d", self.iteration)
+
+            with open(checkpoint, 'rb') as f:
+                self.model_params = pickle.load(f)
+                if self.n_devices > 1:
+                    self.model_params = self.replicate_params(self.model_params)
+        else:
+
+            logger.info("Starting training from scratch")
+            self.model_params = None
+
+    def split_inputs(self, inputs):
+        for k, arr in inputs.items():
+            assert arr.shape[0] % self.n_devices == 0
+            inputs[k] = arr.reshape(
+                self.n_devices, arr.shape[0] // self.n_devices, *arr.shape[1:])
+            inputs[k] = [x for x in inputs[k]]  # make a sequence for put_sharded
+        return inputs
+
+    def unstack_device_outputs(self, outputs):
+        for k, arr in outputs.items():
+            outputs[k] = arr.reshape(arr.shape[0]*arr.shape[1], *arr.shape[2:])
+        return outputs
+
+    def train_step(self, batch, request):
+
+        inputs = self.__collect_provided_inputs(batch)
+        if self.n_devices > 1:
+            inputs = self.split_inputs(inputs)
+
+        # put to device for max performance
+        if self.n_devices > 1:
+            for k, v in inputs.items():
+                inputs[k] = jax.device_put_sharded(v, jax.local_devices())
+        else:
+            for k, v in inputs.items():
+                inputs[k] = jax.device_put(v)
+
+        # initialize model if necessary
+        if self.model_params is None:
+            rng = jax.random.PRNGKey(42)
```

Comment:
Should the random key be hard coded like this?

Reply:
Yea so random key is meant to make training reproducible, but since the Gunpowder pipeline is not deterministic in the first place I guessed it doesn't hurt to hard code the random key for now (and providing an option would be a bit misleading). Can add a comment to make it clearer?

Reply:
Oh ok. It looks like you're targetting the v1.3-dev branch. I built in random seeds to the gunpowder pipeline a little ways back. It might be nice to tie into that. If you look at some of the other nodes that use randomness such as in random location (https://github.com/funkey/gunpowder/blob/4c115bbd944fdc1929427692a9742709c86c12de/gunpowder/nodes/random_location.py#L160) I've been trying to keep random seeds set in the prepare method. This allows requests to be repeated if run with the same seed.
I guess this case is a bit unique since training should change the result of a pipeline request even if the seed used for the request stays the same. But at least if you use the same pipeline with the same initial seed you should get the same first training step.

Reply:
Nice! Yea I can use that `request.random_seed` for the weight init. One step towards deterministic training :)

Reply:
Will the initial seed be the same in the first batch for every run though?

Reply:
The seed can be set when a request is made from the gunpowder pipeline. If the user wants to set the same seed they can, if they want a different seed for multiple random initializations this would make that possible too. I think this looks pretty good now